zookeeper-server/zookeeper-server-3.9.1/src/main/java/org/apache/zookeeper/common/ClientX509Util.java

```diff
@@ -0,0 +1,220 @@
+import org.apache.zookeeper.server.auth.X509AuthenticationProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * <em>NOTE: Overridden because ZK 3.9 completely broke the SSL setup APIs; for clients, key and trust stores are
+ * now mandatory, unlike for servers, where it's still possible to provide a custom authProvider. This patch fixes that.</em>
+ * <p>
+ * X509 utilities specific for client-server communication framework.
+ */
+public class ClientX509Util extends X509Util {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ClientX509Util.class);
+
+    private final String sslAuthProviderProperty = getConfigPrefix() + "authProvider";
+    private final String sslProviderProperty = getConfigPrefix() + "sslProvider";
+
+    @Override
+    protected String getConfigPrefix() {
+        return "zookeeper.ssl.";
+    }
+
+    @Override
+    protected boolean shouldVerifyClientHostname() {
+        return false;
+    }
+
+    public String getSslAuthProviderProperty() {
+        return sslAuthProviderProperty;
+    }
+
+    public String getSslProviderProperty() {
+        return sslProviderProperty;
+    }
+
+    public SslContext createNettySslContextForClient(ZKConfig config)
+        throws X509Exception.KeyManagerException, X509Exception.TrustManagerException, SSLException {
+
+        KeyManager km = null;
+        TrustManager tm = null;
+        String authProviderProp = System.getProperty(getSslAuthProviderProperty());
+        if (authProviderProp == null) {
+            String keyStoreLocation = config.getProperty(getSslKeystoreLocationProperty(), "");
+            String keyStorePassword = getPasswordFromConfigPropertyOrFile(config, getSslKeystorePasswdProperty(),
+                                                                          getSslKeystorePasswdPathProperty());
+            String keyStoreType = config.getProperty(getSslKeystoreTypeProperty());
+            SslContextBuilder sslContextBuilder = SslContextBuilder.forClient();
+            if (keyStoreLocation.isEmpty()) {
+                LOG.warn("{} not specified", getSslKeystoreLocationProperty());
+            } else {
+                km = createKeyManager(keyStoreLocation, keyStorePassword, keyStoreType);
+            }
+            tm = getTrustManager(config);
+        } else {
+            X509AuthenticationProvider authProvider = (X509AuthenticationProvider) ProviderRegistry.getProvider(
+                    System.getProperty(getSslAuthProviderProperty(), "x509"));
+
+            if (authProvider == null) {
+                LOG.error("Auth provider not found: {}", authProviderProp);
+                throw new SSLException("Could not create SSLContext with specified auth provider: " + authProviderProp);
+            }
+            LOG.info("Using auth provider for client: {}", authProviderProp);
+            km = authProvider.getKeyManager();
+            tm = authProvider.getTrustManager();
+        }
+
```

Comment:
This switch is copied from `NettyServerCnxnFactory`; it tries the authProvider, when that is set, instead of jumping straight to file-based key and trust stores. There should be no other changes here, except for a refactoring to avoid some code duplication.